rfcs/convergence/changelog-guidelines.md

```diff
@@ -0,0 +1,130 @@
+# Changelog (beachball changefile) guidelines
+
+## Summary
+
+This RFC proposes a set of guidelines when creating changefiles during a PR. These changefiles are directly translated
+(in most cases) to our changelog which is important for users to see what changed between releases of Fluent. Therefore
+this RFC proposes the following requirements for the changelog:
+
+- Changes that do not affect published code should not be in the changelog
+- Clear labels for change types
+- The detail of changes should be descriptive first and short second
+- Messages should be written for changes specific to the package, not the PR
+
+The changefiles also directly influence how packages are bumped and released in the release pipeline.
+
+## Problem statement
+
+The current v9 changelog contains (or contained before removal) the following bad patterns that should not be in the changelog:
+
+### Changelog entries that do not affect published code
+
+Changes such as:
+
+- stories are updated
+- dev dependencies are updated
+- tests were added/updated
+- scripts were modified
+
+Should not be included in the changelog. Generally when these kinds of changes are in the changelog they should be
+removed as soon as they are noticed.
+
+It's also important that these kinds of change entries **should not bump upstream packages**. By bumping upstream
+packages, we create new releases along the whole chain of dependencies where in fact nothing has actually changed
+for partners. Even in the context of lockstep versioning these kinds of changes should not result in version bumps.
+
+### Changelog entries that are focused on the PR
+
+Quite often, a PR will address/fix an issue that spans several packages. While it's good practice to keep PRs as
+small as possible, it is still possible that mutliple changes in different packages that need to be modified. This
+can also be acceptable when an internal API is modified in a non breaking way, but several consumers need to be updated
+at the same time.
+
+It is very easy to write a single changefile message during PR creating and apply it to all the changed packages in
+the PR. This makes the change harder to understand since the contents might only apply vaguely, or not at all to the package.
+
+A simple example would be `Fix dependencies`, this kind of generic description can be applied to a PR when linked to
+an issue, but has an incredibly vague meaning with applied to the changelog of a package.
+
+## Detailed design or proposal
+
+### Semantic labels for changefile entries
+
+Each changefile entry should have a message that has a semantic label that is inspired by [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
+This RFC proposes four labels in the following form:
+
+```
+feat:  This label communicates that a feature was added to the package
```

Comment:
How do these semantic labels relate to the `type`s in the changefile?

It sounds like the sematic label is for human readers while the type in the changefile is used to drive semver? If this is the case it would be possible to have a message like "fix: fixed a button bug" with a type of `major`?

I ask because you mention taking the semantic names from convetional commits and my experience using conventional commits was with a tool called [commitizen](https://commitizen-tools.github.io/commitizen/) that uses the semantic label in the commit message to drive semver (i.e., `type`).

Have you considered strictly enforcing the semantic label and using it to drive the value of `type` in the changefile?

Reply:
> How do these semantic labels relate to the types in the changefile?

There's almost always a strict correspondence:
```
feat     minor
fix      patch
chore    none
docs     probably none (a bit up to interpretation if it's API docs)
```

> Have you considered strictly enforcing the semantic label and using it to drive the value of type in the changefile?

This is what tools using [conventional changelogs](https://github.com/conventional-changelog/conventional-changelog) do.

We could also consider the inverse (if we're continuing to use beachball with the manual change type prompt): generate the semantic label based on the change type, instead of including it manually.

Reply:
> Have you considered strictly enforcing the semantic label and using it to drive the value of type in the changefile?

It's possible, but likely would involve major changes (perhaps breaking) to our release tool.... and honestly to do that work scares me (you can take a look at the codebase microsoft/beachball)

> generate the semantic label based on the change type

This is actually possible with custom changelog rendering in beachball, the main issue is that as you said. 

Then again the main reason we'd have these conventional commit like labels, is to have some way of categorizing changes in the changelog instead of being a dump of messages

Reply:
> I ask because you mention taking the semantic names from convetional commits and my experience using conventional commits was with a tool called commitizen that uses the semantic label in the commit message to drive semver (i.e., type).

@spmonahan 

Commitizen indeed is amazing and I used it for most of my carer as well. 

The approach that falls short is within a monorepo setup. folks from atlassian created some additional functionality which in the end wasn't solving properly what was needed. Based on multiple lessons learned [changesets](https://github.com/changesets/changesets) was created, which is a de facto "standard" for creating changelogs for OSS monorepos.

Reply:
In the context of our current tooling, and given the 1-1 mapping between change type and semantic label, I still think this would be better handled with beachball custom renderers rather than requiring everyone to manually add appropriate prefixes (and adding danger checks or whatever to enforce it).

The requirement of manually adding prefixes could also create confusion for v8, since in [its changelog](https://github.com/microsoft/fluentui/blob/master/packages/react/CHANGELOG.md) the messages are grouped under major/minor/patch headings and therefore probably shouldn't have additional semantic labels.

Reply:
> beachball custom renderers 

can you provide more context pls?

> The requirement of manually adding prefixes could also create confusion for v8

this RFC is for vNext only correct  @ling1726 ? Question is if we are able to configure beachball for 2 separate workflows

Reply:
For beachball custom renderers, I was originally thinking we could add code to prepend a label in this method:
https://github.com/microsoft/fluentui/blob/580595cf1c42675f9aa6e2f95756d073a2a98a85/scripts/beachball/customRenderers.ts#L34-L41

The problem is the change type isn't actually passed here, so we'd have to either add that on the beachball side or add a custom `renderEntries` (which does get the change type already).

I also realized from looking at the beachball code that once we have multiple change types (not just prerelease), the changelog will go from having a single "Changes" heading to having the change types grouped under headings:
```
### Minor changes
- `pkg1`
  - some minor change

### Patch changes
- `pkg2`
  - some patch change
```

So I'm not sure what would be the preferred approach at that point. Or what the implications are for whether to add any prefix requirements during beta/RC.

> Question is if we are able to configure beachball for 2 separate workflows

The changelong entry has the package name, so it's possible to do something different based on that if needed. (Would want to add a step somewhere to cache the list of v9 packages.)